src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java

```diff
@@ -55,267 +53,360 @@
 import org.apache.lucene.store.IndexInput;
 import org.apache.lucene.util.IOUtils;
 
 /**
- * The {@link IndexDescriptor} is an analog of the SSTable {@link Descriptor} and provides version
- * specific information about the on-disk state of a {@link StorageAttachedIndex}.
- *
- * The {@IndexDescriptor} is primarily responsible for maintaining a view of the on-disk state
- * of an index for a specific {@link org.apache.cassandra.io.sstable.SSTable}. It maintains mappings
+ * The `IndexDescriptor` is an analog of the SSTable {@link Descriptor} and provides version
+ * specific information about the on-disk state of {@link StorageAttachedIndex}es.
+ * <p>
+ * The `IndexDescriptor` is primarily responsible for maintaining a view of the on-disk state
+ * of the SAI indexes for a specific {@link org.apache.cassandra.io.sstable.SSTable}. It maintains mappings
  * of the current on-disk components and files. It is responsible for opening files for use by
  * writers and readers.
- *
- * It's remaining responsibility is to act as a proxy to the {@link OnDiskFormat} associated with the
- * index {@link Version}.
+ * <p>
+ * Each sstable has per-index components ({@link IndexComponent}) associated with it, and also components
+ * that are shared by all indexes (notably, the components that make up the PrimaryKeyMap).
+ * <p>
+ * IndexDescriptor's remaining responsibility is to act as a proxy to the {@link OnDiskFormat}
+ * associated with the index {@link Version}.
  */
 public class IndexDescriptor
 {
     private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
 
-    public final Version version;
+    // TODO Because indexes can be added at any time to existing data, the Version of a column index
+    // may not match the Version of the base sstable.  OnDiskFormat + IndexFeatureSet + IndexDescriptor
+    // was not designed with this in mind, leading to some awkwardness, notably in IFS where some features
+    // are per-sstable (`isRowAware`) and some are per-column (`hasVectorIndexChecksum`).
+
+    // per-SSTable fields
     public final Descriptor descriptor;
     public final IPartitioner partitioner;
     public final ClusteringComparator clusteringComparator;
     public final PrimaryKey.Factory primaryKeyFactory;
-    public final Set<IndexComponent> perSSTableComponents = Sets.newHashSet();
-    public final Map<String, Set<IndexComponent>> perIndexComponents = Maps.newHashMap();
-    public final Map<IndexComponent, File> onDiskPerSSTableFileMap = Maps.newHashMap();
-    public final Map<Pair<IndexComponent, String>, File> onDiskPerIndexFileMap = Maps.newHashMap();
+
+    // versions and components.  null context = per-sstable entry
+    private final Map<IndexContext, Version> versions = Maps.newHashMap();
+    private final Map<IndexContext, Set<IndexComponent>> components = Maps.newHashMap();
+    private final Map<AttachedIndexComponent, File> fileMap = Maps.newHashMap();
```

Comment:
It looks like `IndexDescriptor` is accessed from multiple threads. However, I see that the previous collections present above were not thread safe. Do we have safe publication? I reviewed it a bit and it looks like we probably do.

Reply:
The main "real" place that any of the mutations happen is in the constructor of PerIndexFiles which is called during index initialization.  Everything else is test code.  So (1) it's not worse than it was and (2) spot checking how it's used looks reasonable to me.

I hate the design pattern of "construct an object in one place, initialize it somewhere else" but I didn't want to scope creep this into trying to fix that.

Reply:
Works for me, especially since it follows the existing design pattern and we haven't seen any indications that it is broken.